pkg/kp/rcstore/consul_store.go

```diff
@@ -66,59 +91,42 @@ func (s *consulStore) Create(manifest pods.Manifest, nodeSelector labels.Selecto
 		Disabled:        false,
 	}
 
-	err = s.forEachLabel(rc, func(id, k, v string) error {
-		return s.applicator.SetLabel(labels.RC, id, k, v)
-	})
-
-	// TODO: If the `put` operations fail, we have already labeled the RC,
-	// yet the RC will not exist in the backing Consul KV store.
-	// The labels may have to be cleaned up.
-
-	err = s.put(id, []kvPair{
-		kvPair{key: "disabled", value: []byte("false")},
-		kvPair{key: "node_selector", value: []byte(nodeSelector.String())},
-		kvPair{key: "pod_labels", value: []byte(podLabels.String())},
-		kvPair{key: "pod_manifest", value: buf.Bytes()},
-		kvPair{key: "replicas_desired", value: []byte("0")},
-	})
-
+	jsonRC, err := json.Marshal(rc)
 	if err != nil {
-		return fields.RC{}, err
+		return fields.RC{}, nil
 	}
+	success, _, err := s.kv.CAS(&api.KVPair{
+		Key:   rcp,
+		Value: jsonRC,
+		// the chance of the UUID already existing is vanishingly small, but
+		// technically not impossible, so we should use the CAS index to guard
+		// against duplicate UUIDs
+		ModifyIndex: 0,
+	}, nil)
 
+	if err != nil {
+		return fields.RC{}, kp.NewKVError("cas", rcp, err)
+	}
+	if !success {
+		return fields.RC{}, CASError(rcp)
+	}
 	return rc, nil
 }
 
 func (s *consulStore) Get(id fields.ID) (fields.RC, error) {
-	listed, _, err := s.kv.List(kp.RCPath(id.String()), nil)
-	if err != nil {
+	kvp, _, err := s.kv.Get(kp.RCPath(id.String()), nil)
+	if err != nil || kvp == nil {
 		return fields.RC{}, nil
 	}
-
-	rcMap := s.kvpsToRcs(listed)
-
-	rc, ok := rcMap[id]
-	if !ok {
-		return fields.RC{}, fmt.Errorf("No such replication controller %s", id)
-	}
-
-	return *rc, nil
+	return s.kvpToRC(kvp)
 }
 
 func (s *consulStore) List() ([]fields.RC, error) {
 	listed, _, err := s.kv.List(kp.RC_TREE, nil)
 	if err != nil {
-		return []fields.RC{}, nil
-	}
-
-	rcMap := s.kvpsToRcs(listed)
-	rcs := make([]fields.RC, 0)
-
-	for _, rc := range rcMap {
-		rcs = append(rcs, *rc)
+		return nil, nil
```

Comment:
here also no error is being returned when there is one. was it also unintentional?